### PR TITLE
docs: add feature flags in SvelteKit apps with Neon Postgres

### PR DESCRIPTION
**Pull request by rishi-raj-jain**

### CODE REVIEW

File: src/styles/highlighted-code.css [hunk 1]
```diff
@@ -110,3 +110,11 @@
     }
   }
 }
+
+.diff.add {
```
Comment (saimonkat) [inline src/styles/highlighted-code.css:114]: I'd like to:
- move it up to the `@layer components`
- adjust scss classes
- add before items as `+` and `-`
- update coloes for dark theme support

```
.has-diff code .line {
  @apply pl-6;

  &.diff {
    &.add {
      @apply bg-code-green-1/10 before:absolute before:left-2 before:text-code-green-1 before:content-['+'];
    }
    &.remove {
      @apply bg-code-red-1/10 before:absolute before:left-2 before:text-code-red-1/50 before:content-['-'];
    }
  }
}
```

image

image

File: content/guides/feature-flags-sveltekit.md [hunk 2]
```diff
@@ -0,0 +1,402 @@
+---
+title: Add feature flags in SvelteKit apps with Neon Postgres
+subtitle: Step-by-step guide to integrating feature flags in SvelteKit apps with Postgres
```
Comment (danieltprice) [inline content/guides/feature-flags-sveltekit.md:3]:
```suggestion
subtitle: A step-by-step guide to integrating feature flags in SvelteKit apps with Postgres
```

File: content/guides/feature-flags-sveltekit.md [hunk 3]
```diff
@@ -0,0 +1,402 @@
...
+author: rishi-raj-jain
+enableTableOfContents: true
+createdAt: '2024-05-24T13:24:36.612Z'
+updatedOn: '2024-05-24T13:24:36.612Z'
+---
+
+This guide covers the step-by-step process of integrating feature flags in SvelteKit apps with Postgres (powered by Neon). Feature flags provide a way to control the behavior of your application without deploying new code, allowing you to test and roll out new features dynamically. Upon completing the guide, you would have an understanding of managing and rolling out of new features (such as a hypothetical faster payment method) using dynamic feature flags integration.
```
Comment (danieltprice) [inline content/guides/feature-flags-sveltekit.md:10]:
```suggestion
This guide covers the step-by-step process of integrating feature flags in SvelteKit apps with Postgres (powered by Neon). Feature flags provide a way to control the behavior of your application without deploying new code, allowing you to test and roll out new features dynamically. Upon completing the guide, you will understand how to manage and roll out new features using dynamic feature flag integration.
```

File: content/guides/feature-flags-sveltekit.md [hunk 4]
```diff
@@ -0,0 +1,402 @@
...
+
+## Prerequisites
+
+To follow along this guide, you will need the following:
```
Comment (danieltprice) [inline content/guides/feature-flags-sveltekit.md:14]:
```suggestion
To follow the steps in this guide, you will need the following:
```

File: content/guides/feature-flags-sveltekit.md [hunk 5]
```diff
@@ -0,0 +1,402 @@
...
+
+- [Node.js 18](https://nodejs.org/en/blog/announcements/v18-release-announce) or later
+- A [Neon](https://console.neon.tech/signup) account – The feature flags will be defined (or mutated) in a Serverless Postgres
```
Comment (danieltprice) [inline content/guides/feature-flags-sveltekit.md:17]:
```suggestion
- A [Neon](https://console.neon.tech/signup) account – The feature flags will be defined (or mutated) in a  Postgres database
```

File: content/guides/feature-flags-sveltekit.md [hunk 6]
```diff
@@ -0,0 +1,402 @@
...
+
+## Steps
+
+- [Provisioning a Serverless Postgres powered by Neon](#provisioning-a-serverless-postgres-powered-by-neon)
```
Comment (danieltprice) [inline content/guides/feature-flags-sveltekit.md:21]:
```suggestion
- [Provisioning a Postgres database powered by Neon](#provisioning-a-serverless-postgres-powered-by-neon)
```

File: content/guides/feature-flags-sveltekit.md [hunk 7]
```diff
@@ -0,0 +1,402 @@
...
+- [Create a new SvelteKit application](#create-a-new-sveltekit-application)
```
Comment (danieltprice) [inline content/guides/feature-flags-sveltekit.md:22]:
```suggestion
- [Creating a new SvelteKit application](#create-a-new-sveltekit-application)
```

File: content/guides/feature-flags-sveltekit.md [hunk 8]
```diff
@@ -0,0 +1,402 @@
...
+- [(Optional) Add Tailwind CSS to the application](#optional-add-tailwind-css-to-the-application)
```
Comment (danieltprice) [inline content/guides/feature-flags-sveltekit.md:23]:
```suggestion
- [(Optional) Adding Tailwind CSS to the application](#optional-add-tailwind-css-to-the-application)
```

File: content/guides/feature-flags-sveltekit.md [hunk 9]
```diff
@@ -0,0 +1,402 @@
...
+- [Managing Feature Flags in Serverless Postgres](#managing-feature-flags-in-serverless-postgres)
+- [Dynamic Feature Flag Integration for Testing Fast Payment Methods](#dynamic-feature-flag-integration-for-testing-fast-payment-methods)
+
+## Provisioning a Serverless Postgres powered by Neon
```
Comment (danieltprice) [inline content/guides/feature-flags-sveltekit.md:27]:
```suggestion
## Provisioning a Postgres database powered by Neon
```

File: content/guides/feature-flags-sveltekit.md [hunk 10]
```diff
@@ -0,0 +1,402 @@
...
+- `user` is the database user.
+- `password` is the database user’s password.
+- `endpoint_hostname` is the host with neon.tech as the [TLD](https://www.cloudflare.com/en-gb/learning/dns/top-level-domain/).
+- `port` is the Neon port number. The default port number is 5432.
+- `dbname` is the name of the database. “neondb” is the default database created with each Neon project.
+- `?sslmode=require` an optional query parameter that enforces the [SSL](https://www.cloudflare.com/en-gb/learning/ssl/what-is-ssl/) mode while connecting to the Postgres instance for better security.
+
+Save this connecting string somewhere safe to be used as the `DATABASE_URL` further in the guide. Proceed further in this guide to create a SvelteKit application.
+
+## Create a new SvelteKit application
```
Comment (danieltprice) [inline content/guides/feature-flags-sveltekit.md:52]:
```suggestion
## Creating a new SvelteKit application
```

File: content/guides/feature-flags-sveltekit.md [hunk 11]
```diff
@@ -0,0 +1,402 @@
...
+
+```markdown
+- `@types/pg`: Type definitions for pg.
+- `tsx`: To execute and rebuild TypeScript efficiently.
+- `dotenv`: A library for handling environment variables.
+```
+
+</CodeTabs>
+
+## (Optional) Add Tailwind CSS to the application
```
Comment (danieltprice) [inline content/guides/feature-flags-sveltekit.md:126]:
```suggestion
## (Optional) Adding Tailwind CSS to the application
```

File: content/guides/feature-flags-sveltekit.md [hunk 12]
```diff
@@ -0,0 +1,402 @@
...
+
+```bash
+npm run dev
+```
+
+The app now should be running on [localhost:5173](http://localhost:5173).
+
+> Note: According to an [advanced SvelteKit guide](https://kit.svelte.dev/docs/server-only-modules), using `.server` in the filename allows you to mark the code to be executed on server only.
+
+Next, run the command below to install the necessary libraries and packages for building the application:
```
Comment (danieltprice) [inline content/guides/feature-flags-sveltekit.md:75]:
```suggestion
Next, run the commands below to install the necessary libraries and packages for building the application:
```

File: content/guides/feature-flags-sveltekit.md [hunk 13]
```diff
@@ -0,0 +1,402 @@
...
+```
+
+```bash
+npm install pg
+npm install -D dotenv tsx @types/pg
+```
+
+</CodeTabs>
+
+It installs the packages passed to the install command, with the -D flag specifying the libraries intended for development purposes only.
```
Comment (danieltprice) [inline content/guides/feature-flags-sveltekit.md:91]:
```suggestion
The commands install the required libraries and packages, with the `-D` flag specifying the libraries intended for development purposes only.
```

File: content/guides/feature-flags-sveltekit.md [hunk 14]
```diff
@@ -0,0 +1,402 @@
...
+
+
+
+The libraries installed include:
+
+<CodeTabs labels={["Neon serverless driver", "node-postgres"]}>
+
+```markdown
+- `@neondatabase/serverless`: Neon's PostgreSQL driver for JavaScript and TypeScript.
```
Comment (danieltprice) [inline content/guides/feature-flags-sveltekit.md:100]:
```suggestion
- `@neondatabase/serverless`: Neon's serverless Postgres driver for JavaScript and TypeScript.
```

File: content/guides/feature-flags-sveltekit.md [hunk 15]
```diff
@@ -0,0 +1,402 @@
...
+```
+
+```markdown
+- `pg`: A PostgreSQL client for Node.js.
```
Comment (danieltprice) [inline content/guides/feature-flags-sveltekit.md:104]:
```suggestion
- `pg`: A Postgres client for Node.js.
```

File: content/guides/feature-flags-sveltekit.md [hunk 16]
```diff
@@ -0,0 +1,402 @@
...
+```
+
+</CodeTabs>
+
+The development-specific libraries include:
+
+<CodeTabs labels={["Neon serverless driver", "node-postgres"]}>
+
+```markdown
+- `tsx`: To execute and rebuild TypeScript efficiently.
```
Comment (danieltprice) [inline content/guides/feature-flags-sveltekit.md:114]:
```suggestion
- `tsx`: A library for executing and rebuilding TypeScript efficiently.
```

File: content/guides/feature-flags-sveltekit.md [hunk 17]
```diff
@@ -0,0 +1,402 @@
...
+- `dotenv`: A library for handling environment variables.
+```
+
+```markdown
+- `@types/pg`: Type definitions for pg.
+- `tsx`: To execute and rebuild TypeScript efficiently.
```
Comment (danieltprice) [inline content/guides/feature-flags-sveltekit.md:120]:
```suggestion
- `tsx`: A library for executing and rebuilding TypeScript efficiently.
```

File: content/guides/feature-flags-sveltekit.md [hunk 18]
```diff
@@ -0,0 +1,402 @@
...
+- `dotenv`: A library for handling environment variables.
+```
+
+</CodeTabs>
+
+## (Optional) Add Tailwind CSS to the application
+
+For styling the app, we will be using Tailwind CSS. Install and set up Tailwind at the root of our project's directory by running:
```
Comment (danieltprice) [inline content/guides/feature-flags-sveltekit.md:128]:
```suggestion
For styling the app, we will use Tailwind CSS. Install and set up Tailwind at the root of our project's directory by running:
```

File: content/guides/feature-flags-sveltekit.md [hunk 19]
```diff
@@ -0,0 +1,402 @@
...
+
+```bash
+npm install -D tailwindcss postcss autoprefixer
+npx tailwindcss init -p
+```
+
+Create an `app.css` file in the `src` directory, and add the snippet below in it:
```
Comment (danieltprice) [inline content/guides/feature-flags-sveltekit.md:135]:
```suggestion
Create an `app.css` file in the `src` directory, and add the snippet below:
```

File: content/guides/feature-flags-sveltekit.md [hunk 20]
```diff
@@ -0,0 +1,402 @@
...
+</script> // [!code ++]
+
+<!-- +page.svelte's HTML -->
+```
+
+## Managing Feature Flags in Serverless Postgres
+
+Feature flags offer a powerful way to control the behavior of your application without deploying new code. In a Serverless Postgres environment, you can easily create, read, and update feature flags using the following steps:
+
+### Create a Serverless Postgres Client
```
Comment (danieltprice) [inline content/guides/feature-flags-sveltekit.md:171]:
```suggestion
### Create a serverless Postgres client
```

File: content/guides/feature-flags-sveltekit.md [hunk 21]
```diff
@@ -0,0 +1,402 @@
...
+// Create a connection string to the Neon Postgres instance
+const connectionString: string = process.env.DATABASE_URL as string
+
+// Create a in-memory pool so that it's cached for multiple calls
+export default new pg.Pool({ connectionString })
+```
+
+</CodeTabs>
+
+The code above starts with importing the Postgres client. It further imports the config module by `dotenv` that makes sure that all environment variables are populated in application environment. It then creates a new instance of a PostgreSQL connection pool.
```
Comment (danieltprice) [inline content/guides/feature-flags-sveltekit.md:211]:
```suggestion
The code above starts with importing the Postgres client. It further imports the config module by `dotenv` that makes sure that all environment variables are populated in application environment. It then creates a new instance of a Postgres connection pool.
```

File: content/guides/feature-flags-sveltekit.md [hunk 22]
```diff
@@ -0,0 +1,402 @@
...
+  })
+  console.log('✅ Setup an enabled feature flag to accept fast payment methods.')
+}
+
+populateFeatureFlags()
+```
+
+</CodeTabs>
+
+The code snippet above first ensures the existence of a table named `feature_flags` in the PostgreSQL database. Then, it inserts a feature flag named `fast_payments` with the value `true`, indicating that fast payment methods are enabled.
```
Comment (danieltprice) [inline content/guides/feature-flags-sveltekit.md:260]:
```suggestion
The code snippet above first ensures the existence of a table named `feature_flags` in the Postgres database. Then, it inserts a feature flag named `fast_payments` with the value `true`, indicating that fast payment methods are enabled.
```

File: content/guides/feature-flags-sveltekit.md [hunk 23]
```diff
@@ -0,0 +1,402 @@
...
+
+### READ and UPDATE the Feature Flags
```
Comment (danieltprice) [inline content/guides/feature-flags-sveltekit.md:262]:
```suggestion
### Read and update the feature flags
```

File: content/guides/feature-flags-sveltekit.md [hunk 24]
```diff
@@ -0,0 +1,402 @@
...
+}
+```
+
+</CodeTabs>
+
+The `setEnabled` function updates the value of a feature flag in the database. In this example, you will update the `fast_payments` feature dynamically per request to get a taste of how feature flags are used in production.
+
+Great! You can use these helper functions in your application to manage and control feature flags dynamically.
+
+## Dynamic Feature Flag Integration for Testing Fast Payment Methods
```
Comment (danieltprice) [inline content/guides/feature-flags-sveltekit.md:330]:
```suggestion
## Dynamic feature flag integration for testing fast payment methods
```

File: content/guides/feature-flags-sveltekit.md [hunk 25]
```diff
@@ -0,0 +1,402 @@
...
+
+In this section, you will get an example of how feature flag helps test and roll out new features, dynamically. For example, you are a payment processing company. You have just added a payment method named `PayGM` that helps users pay faster. But you want to test it out on a random basis for each cart that you process. Let's walk through a hypothetical code that allows you to understand the usage of feature flag in this case.
```
Comment (danieltprice) [inline content/guides/feature-flags-sveltekit.md:332]:
```suggestion
In this section, you will get an example of how a feature flag helps test and roll out new features, dynamically. For example, you are a payment processing company. You have just added a payment method named `PayGM` that helps users pay faster. But you want to test it out on a random basis for each cart that you process. Let's walk through the hypothetical code to understand the usage of a feature flag in this case.
```

File: content/guides/feature-flags-sveltekit.md [hunk 26]
```diff
@@ -0,0 +1,402 @@
...
+
+### Computing the user destination
+
+In a SvelteKit route, the data from the server to the user interface is passed via `+page.server.ts` file to `+page.svelte`. For the sake of an example, you will load the feature flag dynamically and check if it's enabled to determine the user's destination experience. To do that, add the following snippet to `+page.server.ts` file:
```
Comment (danieltprice) [inline content/guides/feature-flags-sveltekit.md:336]:
```suggestion
In a SvelteKit route, the data from the server to the user interface is passed via `+page.server.ts` file to `+page.svelte`. For the sake of this example, you will load the feature flag dynamically and check if it's enabled to determine the user's destination experience. To do that, add the following snippet to `+page.server.ts` file:
```

File: content/guides/feature-flags-sveltekit.md [hunk 27]
```diff
@@ -0,0 +1,402 @@
...
+    else cookies.set('destination_bucket', '0', { path: '/' })
+  }
+  const fast_payments = Boolean(Number(cookies.get('destination_bucket')))
+  return {
+    fast_payments,
+  }
+}
+```
+
+The code above first looks for the bucket assigned to the user. If no such bucket is found, it looks for the value of the feature flag in the database, randomly assigns a boolean whenever the `/` route is visited and sets it into the cookie. Finally, it reads the cookie as the source to determine the user experience, in here to check if the fast payment methods are enabled or not.
```
Comment (danieltprice) [inline content/guides/feature-flags-sveltekit.md:360]:
```suggestion
The code above first looks for the bucket assigned to the user. If no such bucket is found, it looks for the value of the feature flag in the database, randomly assigns a boolean whenever the `/` route is visited, and sets the value in the cookie. Finally, it reads the cookie as the source to determine the user experience and check if the fast payment methods are enabled or not.
```

File: content/guides/feature-flags-sveltekit.md [hunk 28]
```diff
@@ -0,0 +1,402 @@
...
+
+### Creating a Conditional User Experience
```
Comment (danieltprice) [inline content/guides/feature-flags-sveltekit.md:362]:
```suggestion
### Creating a conditional user experience
```

**Review by danieltprice — APPROVED** [review]
A few minor suggestions + a couple of style revisions:
- Use sentence-style capitalization
- Use "Postgres" instead of "PostgreSQL"

File: src/styles/highlighted-code.css (hunk 1 above)
Comment (rishi-raj-jain) [reply src/styles/highlighted-code.css]: Done by yourself.

**Comment by rishi-raj-jain** [comment]
@danieltprice 

Re-requesting your review for the source code block that I have added.

**Comment by rishi-raj-jain** [comment]
Preview: https://neon-next-e4n8aenox-neondatabase.vercel.app/guides/feature-flags-sveltekit.

**Pull request merged**